jms/src/main/scala/akka/stream/alpakka/jms/JmsConnector.scala

```diff
@@ -48,36 +50,121 @@ private[jms] trait JmsConnector[S <: JmsSession] { this: GraphStageLogic =>
+        case Some(Credentials(username, password)) => factory.createConnection(username, password)
+        case _ => factory.createConnection()
+      }
+      if (status.get == Connecting) { // `TimedOut` can be set at any point. So we have to check whether to continue.
+        connectionRef.set(Some(connection))
+        if (startConnection) connection.start()
+      }
+      // ... and close if the connection is not to be used, don't return the connection
+      if (!status.compareAndSet(Connecting, Connected)) {
+        connectionRef.get.foreach(_.close())
+        connectionRef.set(None)
+        throw new TimeoutException("Received timed out signal trying to establish connection")
+      } else connection
+    }
+
+    val connectTimeout = jmsSettings.connectionRetrySettings.connectTimeout
+    val timeoutFuture = after(connectTimeout, system.scheduler) {
+      // Even if the timer goes off, the connection may already be good. We use the
+      // status field and an atomic compareAndSet to see whether we should indeed time out, or just return
+      // the connection. In this case it does not matter which future returns. Both will have the right answer.
+      if (status.compareAndSet(Connecting, TimedOut)) {
+        connectionRef.get.foreach(_.close())
+        connectionRef.set(None)
+        Future.failed(
+          new TimeoutException(
+            s"Timed out after $connectTimeout trying to establish connection. " +
+            "Please see ConnectionRetrySettings.connectTimeout"
+          )
+        )
+      } else
+        connectionRef.get match {
+          case Some(connection) => Future.successful(connection)
+          case None => Future.failed(new IllegalStateException("BUG: Connection reference not set when connected"))
+        }
     }
-    connection.setExceptionListener(new jms.ExceptionListener {
-      override def onException(exception: jms.JMSException): Unit =
-        fail.invoke(exception)
-    })
-    onConnection.invoke(connection)
-    connection
+
+    Future.firstCompletedOf(Iterator(connectionFuture, timeoutFuture))(ExecutionContexts.sameThreadExecutionContext)
+  }
+
+  private def openConnectionWithRetry(startConnection: Boolean, n: Int = 0, maxed: Boolean = false)(
+      implicit system: ActorSystem
+  ): Future[jms.Connection] =
+    openConnection(startConnection).recoverWith {
+      case e: jms.JMSSecurityException => Future.failed(e)
+      case NonFatal(t) =>
+        val retrySettings = jmsSettings.connectionRetrySettings
+        import retrySettings._
+        val nextN = n + 1
+        if (maxRetries >= 0 && nextN > maxRetries) { // Negative maxRetries treated as infinite.
+          if (maxRetries == 0) Future.failed(t)
+          else Future.failed(ConnectionRetryException(s"Could not establish connection after $n retries.", t))
+        } else {
+          val delay = if (maxed) maxBackoff else waitTime(nextN)
+          if (delay >= maxBackoff) {
```

Comment:
why not just
```scala
          val delay = waitTime(nextN).min(maxBackoff)
          after(delay, system.scheduler) {
            openConnectionWithRetry(startConnection, nextN)
          }
```
? :)

Reply:
Since `waitTime` is an exponential function, there is a very small chance a very large backoff may overflow, throwing this logic into haywire. That's why I really want to stop calculating the `waitTime` after max is reached.